ankimorphs/highlight_morphs_jit.py

```diff
@@ -0,0 +1,149 @@
+from .ankimorphs_db import AnkiMorphsDB
+from .morpheme import Morpheme
+from .morphemizers import morphemizer as morphemizer_module
+from .morphemizers import spacy_wrapper
+from .morphemizers.morphemizer import Morphemizer, SpacyMorphemizer
+
+
+def highlight_morphs_jit(
+    field_text: str,
+    field_name: str,
+    filter_name: str,
+    context: TemplateRenderContext,
+) -> str:
+    """Use morph learning progress to decorate the morphemes in the supplied text.
+    Adds css classes to the output that can be styled in the card."""
+
+    # Perf: Bail early if the user attempts to use this template filter on the already
+    # formatted data.
+    #
+    if (
+        filter_name != "am-highlight"
+        or field_name == ankimorphs_globals.EXTRA_FIELD_HIGHLIGHTED
+    ):
+        return field_text
+
+    am_config_filter: AnkiMorphsConfigFilter | None = (
+        ankimorphs_config.get_matching_filter(context.note())
+    )
+
+    if am_config_filter is None:
+        return field_text
+
+    morphemizer: Morphemizer | None = morphemizer_module.get_morphemizer_by_description(
+        am_config_filter.morphemizer_description
+    )
+
+    if not morphemizer:
+        return field_text
+
+    am_config = AnkiMorphsConfig()
+
+    card_morphs: list[Morpheme] = _get_morph_meta_for_text(
+        morphemizer, field_text, am_config
+    )
+
+    if not card_morphs:
+        return field_text
+
+    return text_highlighting.alt_get_highlighted_text(
+        am_config,
+        card_morphs,
+        _dehtml(field_text),
+    )
+
+
+def _get_morph_meta_for_text(
+    morphemizer: Morphemizer,
+    field_text: str,
+    am_config: AnkiMorphsConfig,
+) -> list[Morpheme]:
+    """Take in a string and gather the morphemes from it."""
+
+    # If we were piped in after the `furigana` built-in filter, or if there is html in the source
+    # data, we need to do some cleansing.
+    #
+    clean_text = _dehtml(field_text, am_config, True)
+
+    if isinstance(morphemizer, SpacyMorphemizer):
+        nlp = spacy_wrapper.get_nlp(
+            morphemizer.get_description().removeprefix("spaCy: ")
+        )
+
+        morphs = text_preprocessing.get_processed_spacy_morphs(
+            am_config, next(nlp.pipe([clean_text]))
+        )
+    else:
+        morphs = text_preprocessing.get_processed_morphemizer_morphs(
+            morphemizer, clean_text, am_config
+        )
+
+    morphs = list(set(morphs))
+
+    if not morphs:
+        return []
+
+    with AnkiMorphsDB() as am_db:
+        for morph in morphs:
+            if am_config.evaluate_morph_inflection:
+                morph.highest_inflection_learning_interval = (
+                    am_db.get_highest_inflection_learning_interval(morph) or 0
+                )
+            else:
+                morph.highest_lemma_learning_interval = (
+                    am_db.get_highest_lemma_learning_interval(morph) or 0
+                )
+
+    return morphs
```

Comment:
Because we're called dynamically, we need to look up the morphs and get the data about them manually.